File: anagrams.py
"""
Alice is taking a cryptography class and finding anagrams to be very useful.
We consider two strings to be anagrams of each other if the first string's letters
can be rearranged to form the second string. In other words, both strings must
contain the same exact letters in the same exact frequency
For example, bacdc and dcbac are anagrams, but bacdc and dcbad are not.

Alice decides on an encryption scheme involving two large strings where encryption
is dependent on the minimum number of character deletions required to make the two
strings anagrams. Can you help her find this number?

Given two strings,  and , that may or may not be of the same length, determine the
minimum number of character deletions required to make  and  anagrams. Any characters
can be deleted from either of the strings.
url: https://www.hackerrank.com/challenges/ctci-making-anagrams
"""


def number_needed(a,b):
    """
        counts the 'minimum' amount of character that needs to be removed from 2 strings
        in order to get them anagrams of each other
    :param a: first string to be counted
    :param b: second string to be counted
    :return: total amount of strings that need to be removed
    """
    count = 0
    for x in set(a):
        count += min(a.count(x), b.count(x))

    return len(a) + len(b) - 2*count

def case0():
    a = "abc"
    b = "cde"
    return number_needed(a, b)

def case1():
    a = "gwoydkkstkgaluglmwusqlpgozlvocxskfrrfhlowwzybguzps"
    b = "qlkwlpwhbtuefedscyeualrzfdnzks"
    return number_needed(a, b)

def case2():
    a="afaaa"
    b="afff"
    return number_needed(a, b)

def case3():
    a="afaaa"
    b="afff"
    return number_needed(a, b)

def case4():
    a="bacdc"
    b="dcbac"
    return number_needed(a, b)

def case5():
    a="bacdc"
    b="dcbad"
    return number_needed(a, b)

assert case0() == 4
assert case1() == 30
assert case2() == 5
assert case4() == 0
assert case5() == 2

File: hackerrank/cpp/anagrams.cc
#include <string>
#include <iostream>
#include <cstdio>
#include <map>
#include <algorithm>
#include <cassert>

using namespace std;

int countString(string complete, const char tofind){
  int count = 0;
  size_t nPos = complete.find(tofind, 0);
  while(nPos != string::npos){
    count ++;
    nPos = complete.find(tofind,nPos+1);
  }
  return count;
}

map<char,char> toMap(string complete){
  std::map<char,char> mymap;
  for (string::iterator it = complete.begin(); it != complete.end(); ++it) {
    mymap.insert ( std::pair<char,char>(*it,*it) );
  }
  return mymap;
}

int number_needed(string a, string b) {
  map<char, char> result = toMap(a);
  int count = 0;
  for (map<char,char>::iterator it = result.begin(); it != result.end(); ++it) {
    count += min(int(countString(a, it->first)), int(countString(b, it->first)));
  }
  return (a.length() + b.length()) - 2*count;
}

void case0(){
  string a = "abc";
  string b = "cde";

  int n = number_needed(a, b);
  printf("\nresult: %d\n", n);
  assert(n == 4);
}

void case1(){
  string a = "gwoydkkstkgaluglmwusqlpgozlvocxskfrrfhlowwzybguzps";
  string b = "qlkwlpwhbtuefedscyeualrzfdnzks";

  int n = number_needed(a, b);
  printf("\nresult: %d\n", n);
  assert(n == 30);
}

void case2(){
  string a="afaaa";
  string b="afff";

  int n = number_needed(a, b);
  printf("\nresult: %d\n", n);
  assert(n == 5);
}

void case3(){
  string a="bacdc";
  string b="dcbad";

  int n = number_needed(a, b);
  printf("\nresult: %d\n", n);
  assert(n == 2);
}

int main(){
  case0();
  case1();
  case2();
  case3();
}

int main_from_test(){
    string a;
    cin >> a;
    string b;
    cin >> b;
    cout << number_needed(a, b) << endl;
    return 0;
}
